### PR TITLE
New Tests and Contract Update

### PR DESCRIPTION
**Pull request by remedcu**
## TLDR
- Updated the contract to have view functions for reading private storage.
- Added more tests for higher coverage of all possible code paths.
- Updated the storage of contracts to be `private` to keep consistency.

## LLM Description
This pull request introduces several improvements and refactorings to the policy engine contracts and their associated tests. The main themes are: enhanced encapsulation of internal mappings, addition of public view functions for policy state inspection, improved error handling, and expanded test coverage for policy edge cases.

**Smart contract improvements and refactoring:**

* Internal mappings in `AllowedModulePolicy`, `CoSignerPolicy`, `ERC20ApprovePolicy`, and `ERC20TransferPolicy` are now private, using the ` prefix for clarity and encapsulation. Public view functions (`isModuleAllowed`, `getCoSigner`, `isSpenderAllowed`, `isRecipientAllowed`) have been added to allow external inspection of policy state without exposing the mappings directly. [[1]](diffhunk://#diff-fe87765f9b30bc26c075bf4e826057bc7dfffd86b0639f4a96e56fe22d135027L14-R16) [[2]](diffhunk://#diff-fe87765f9b30bc26c075bf4e826057bc7dfffd86b0639f4a96e56fe22d135027L60-R69) [[3]](diffhunk://#diff-a0b084b6c0208030911b624e1924216f38bdad5db0572328c1196518e1141dddR14-R23) [[4]](diffhunk://#diff-a0b084b6c0208030911b624e1924216f38bdad5db0572328c1196518e1141dddL59-R77) [[5]](diffhunk://#diff-3b326ea661d61309fed63bb199ad6380d0c579b4c7d5516b5d011d72e8492802R25-R27) [[6]](diffhunk://#diff-3b326ea661d61309fed63bb199ad6380d0c579b4c7d5516b5d011d72e8492802R95-R111) [[7]](diffhunk://#diff-22325b21959761caad1a5ec945f6b94e12054e9ba67bbe02dfbddccd1fd38283R25-R27) [[8]](diffhunk://#diff-22325b21959761caad1a5ec945f6b94e12054e9ba67bbe02dfbddccd1fd38283R99-R115)

* The `PolicyEngine` contract's `confirmPolicy` method now configures the policy before updating the mapping, ensuring atomicity and correctness in policy application.

* Improved error handling in `CoSignerPolicy` by adding a specific revert for the case where no cosigner is configured.

**Testing and documentation:**

* Added a comprehensive test suite for `AllowedModulePolicy`, including both unit and integration tests with `SafePolicyGuard`, covering allowed and unauthorized modules.

* Expanded the `ERC20TransferPolicy` test suite to cover edge cases such as empty recipient list configuration and toggling recipient permissions.

* Updated the `README.md` with clear instructions for testing, gas benchmarking, and deployment, as well as a note on configuring delay parameters.

**Test improvements:**

* Refactored test fixtures to include all relevant signers and deployed contracts, and added an `accessSelector` for more robust policy configuration in tests. [[1]](diffhunk://#diff-4d550ab6f80206c8631c9ce458e67cbcb82d6a0c066867420bcd786888a8d685L11-R11) [[2]](diffhunk://#diff-4d550ab6f80206c8631c9ce458e67cbcb82d6a0c066867420bcd786888a8d685R35-R48)

These changes collectively improve code maintainability, test coverage, and usability for developers and auditors.

### CODE REVIEW

File: contracts/core/PolicyEngine.sol [hunk 1]
```diff
@@ -154,11 +154,13 @@ abstract contract PolicyEngine is IPolicyEngine {
     ) internal {
         // Creating access selector for a policy
         AccessSelector.T access = AccessSelector.create(target, selector, operation);
-        $policies[safe][access] = policy;
 
         // Configuring policy
         require(IPolicy(policy).configure(safe, access, data), PolicyConfigurationFailed());
 
+        // Update the policy mapping
+        $policies[safe][access] = policy;
```
Comment (nlordell) [inline contracts/core/PolicyEngine.sol:162]: There is a potential reentrency issue here right? Shouldn't we set storage before calling the policy?
